cloud-controller-manager/do/loadbalancers.go

```diff
@@ -388,16 +388,26 @@ func buildHealthCheck(service *v1.Service) (*godo.HealthCheck, error) {
 	if err != nil {
 		return nil, err
 	}
+	port := service.Spec.Ports[0]
 
 	// if no health check protocol is specified, use the protocol used for
-	// load balancer traffic.
+	// load balancer traffic, unless TLS passthrough is enabled, then default
+	// to the protocol of the health check port.
 	if healthCheckProtocol == "" {
 		protocol, err := getProtocol(service)
 		if err != nil {
 			return nil, err
 		}
 
-		healthCheckProtocol = protocol
+		tlsPassThrough := getTLSPassThrough(service)
+
+		if tlsPassThrough || (protocol != "https") {
+			healthCheckProtocol = protocol
```

Comment:
We are going to set the health check protocol to `https` here If `tlsPassThrough` is true and `protocol` happens to be _https_. However, the [Load Balancer API supports "http" and "tcp" only](https://developers.digitalocean.com/documentation/v2/#load-balancers) (which is also what we validate in our health check protocol getter).

I am thinking that we could simplify the entire logic as following:

- If `healthCheckProtocol` is set: use it.
- If `healthCheckProtocol` is _not_ set: use TCP

While we could try to be smart and guess if the user wants TCP or HTTP as the health check protocol based on what the TLS pass-through and protocol values look like, I don't think we can really make any reliable assumptions in this case: say that TLS pass-through is **not** set and protocol is set to HTTPS, does that really mean that the user wants to have HTTP set up as the health check protocol? If the application does not implement HTTP verb semantics correctly, even HTTP GET requests for health checks could mutate the state within the application in undesirable ways. (That's a big no-go you shouldn't do, but sometimes you can't easily do the right thing if you're dealing with third-party application.)

My gut feeling is we should stick to the most simple way that is still helpful, which is to fall back to TCP. This would mean a breaking change in behavior but I think it's the better option than taking fuzzy assumptions; even more so when HTTP/2 becomes supported.

WDYT?

Reply:
That makes a lot of sense to me. I pushed an update to that effect :+1: